src/services/ftp/backend.rs

```diff
@@ -0,0 +1,440 @@
+
+        // change to the root path
+        ftp_stream.cwd(&root).map_err(|e| {
+            other(BackendError::new(
+                HashMap::from([("root".to_string(), root.clone())]),
+                anyhow!(e),
+            ))
+        })?;
+
+        let client = ftp_stream;
+
+        info!("ftp backend finished: {:?}", &self);
+        Ok(Backend {
+            endpoint: endpoint.to_string(),
+            root,
+            client: Arc::new(Mutex::new(client)),
+            enable_secure,
+        })
+    }
+}
+
+/// Backend is used to serve `Accessor` support for ftp.
+#[derive(Clone)]
+pub struct Backend {
+    endpoint: String,
+    root: String,
+    client: Arc<Mutex<FtpStream>>,
+    enable_secure: bool,
+}
+
+impl Debug for Backend {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Backend")
+            .field("endpoint", &self.endpoint)
+            .field("root", &self.root)
+            .field("tls", &self.enable_secure)
+            .finish()
+    }
+}
+
+impl Backend {
+    pub(crate) fn from_iter(it: impl Iterator<Item = (String, String)>) -> Result<Self> {
+        let mut builder = Builder::default();
+
+        for (k, v) in it {
+            let v = v.as_str();
+            match k.as_ref() {
+                "root" => builder.root(v),
+                _ => continue,
+            };
+        }
+        builder.build()
+    }
+}
+
+#[async_trait]
+impl Accessor for Backend {
+    async fn create(&self, args: &OpCreate) -> Result<()> {
+        let path = args.path();
+        if args.mode() == ObjectMode::FILE {
+            //let mut ftp_stream = self.ftp_connect()?;
+            self.client
+                .try_lock()
+                .unwrap()
+                .put_file(&path, &mut "".as_bytes())
+                .map_err(|e| {
+                    other(ObjectError::new(
+                        Operation::Create,
+                        path,
+                        anyhow!("put request: {e:?}"),
+                    ))
+                })?;
+            return Ok(());
+        }
+
+        if args.mode() == ObjectMode::DIR {
+            self.client.try_lock().unwrap().mkdir(&path).map_err(|e| {
+                other(ObjectError::new(
+                    Operation::Create,
+                    path,
+                    anyhow!("mkdir request: {e:?}"),
+                ))
+            })?;
+
+            return Ok(());
+        }
+
+        unreachable!()
+    }
+
+    async fn read(&self, args: &OpRead) -> Result<BytesReader> {
+        let path = args.path();
+        let mut guard = self.client.lock().await;
+
+        let mut stream = guard.retr_as_stream(path).map_err(|e| {
+            other(ObjectError::new(
+                Operation::Read,
+                path,
+                anyhow!("retrieve request: {e:?}"),
+            ))
+        })?;
+
+        if let Some(offset) = args.offset() {
+            copy(&mut stream.by_ref().take(offset), &mut sink())?;
```

Comment:
Maybe we can use https://docs.rs/suppaftp/latest/suppaftp/struct.FtpStream.html#method.resume_transfer

Reply:
```
[2022-09-03T06:15:25Z DEBUG suppaftp::sync_ftp] PASV command
Error: object error: (op: read, path: 76c4462b-9b26-4102-9d38-e27e155b361b, source: retrieve request: UnexpectedResponse(Response { status: FileUnavailable, body: "550 No support for resume of ASCII transfer." }))

```
The `resume_transfer()` function seems only support for binary transfer.

Reply:
Of course, we must use binary transfer.